### PR TITLE
Implement Keepalive

### LINKED ISSUE #132: Implement Keepalive Connection String Options
`Keep Alive, Keepalive` (default: 0) For TCP connections, idle connection time measured in seconds, before the first keepalive packet is sent. A value of 0 indicates that keepalive is not used.

Should call `socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, value)` on TCP Socket if `value > 0`

### PR DESCRIPTION
**Pull request by caleblloyd**
- Fixes #132 
- A value of 0 indicates that the OS Default keepalive settings are used
- On Windows, a value greater than 0 is the idle connection time, measured in seconds, before the first keepalive packet is sent.
- Unix-based Operating Systems will always use the OS Default keepalive settings.

### CODE REVIEW

**Comment by caleblloyd** [comment]
I've included a basic test, but I couldn't think of a better way to unit test this since everything happens at the network driver layer.  Here's a screenshot of a packet capture running this commit with `Keepalive=3` in the connection string.

keepalive

As you can see, the keepalive runs every 3 seconds from time 792-810.  At time 812, I shutdown the MySql server, and the server sends FIN.

The client does not send FIN, so the client believes the connection is still open.  The next keepalive packet is not ACK'd, so the client retries keepalives every 1s, which is the proper interval.  Eventually 10 keepalives are missed, so the client RST the connection.

I think that this demonstrates correct behavior.

File: tests/SideBySide.New/ConnectAsync.cs [hunk 1]
```diff
@@ -101,6 +103,21 @@ public ConnectAsync(DatabaseFixture database)
...
+		{
+			var csb = AppConfig.CreateConnectionStringBuilder();
+			csb.Keepalive = 1;
+			using (var connection = new MySqlConnection(csb.ConnectionString))
+			{
+				await connection.OpenAsync();
+				var stopwatch = Stopwatch.StartNew();
+				await Task.Delay(3000);
+				stopwatch.Stop();
+				Assert.InRange(stopwatch.ElapsedMilliseconds, 2500, 3500);
```
Comment (bgrainger) [inline tests/SideBySide.New/ConnectAsync.cs]: What is this testing specifically? That no exception is thrown?

File: docs/content/connection-options.md [hunk 2]
```diff
@@ -148,6 +148,13 @@ These are the other options that MySqlConnector supports.  They are set to sensi
     <td>True to have MySqlDataReader.GetValue() and MySqlDataReader.GetDateTime() return DateTime.MinValue for date or datetime columns that have disallowed values.</td>
   </tr>
   <tr>
+    <td>Keep Alive, Keepalive</td>
+    <td>0</td>
+    <td>TCP Keepalive idle time.  A value of 0 indicates that the OS Default keepalive settings are used.
```
Comment (bgrainger) [inline docs/content/connection-options.md]: Connector/Net documents "A value of 0 indicates that keepalive is not used."; do we want to be different by default?

File: src/MySqlConnector/Serialization/SerializationUtility.cs [hunk 3]
```diff
@@ -47,5 +51,30 @@ public static int GetBytesPerCharacter(CharacterSet characterSet)
 				return 1;
 			}
 		}
+
+		public static void SetKeepalive(Socket socket, uint keepAliveTimeSeconds)
+		{
+			// Always use the OS Default Keepalive settings
+			socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
```
Comment (bgrainger) [inline src/MySqlConnector/Serialization/SerializationUtility.cs]: Following up on question above, should this option only be set if user has set keepAlive != 0?

If you think that the benefits of turning it on by default are worthwhile, then I have no objection. (I'm in favour of changing to use sensible defaults.)

File: tests/SideBySide.New/ConnectAsync.cs (hunk 1 above)
Comment (caleblloyd) [inline tests/SideBySide.New/ConnectAsync.cs]: Correct.  I guess the stopwatch is not necessary, I'll just remove that

File: src/MySqlConnector/Serialization/SerializationUtility.cs (hunk 3 above)
Comment (caleblloyd) [inline src/MySqlConnector/Serialization/SerializationUtility.cs]: I think that always enabling it to use the OS Default would be sensible.  It allows the OS to clean up old sockets should the connection go away.  If data is consistently being sent across the network, Keepalive packets will never need to be sent.

[Windows](https://blogs.technet.microsoft.com/nettracer/2010/06/03/things-that-you-may-want-to-know-about-tcp-keepalives/) and [Linux](http://www.tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/#configuringkernel) both default to send the first Keepalive packet after 2 hours of inactivity on the socket.

**Pull request merged**